### ISSUE TITLE
VirtualScheme review changes

### ISSUE THREAD

**Issue by HaoK** [aspnet/Security#1546]
Per review with @davidfowl @Tratcher @DamianEdwards @ajcvickers 

- Roll VirtualScheme functionality into base AuthenticationSchemeOptions
- Add new top level methods that combine the social OAuth/OIDC auths and cookies with default configuration into `AddAuthentication().AddGoogle/FacebookScheme()` that will take care of everything needed (ensuring only a single cookie is configured, and setting SignInSchemes)

**Comment by HaoK** [comment]
I did a quick prototype of what a higher level "Add[Social]Scheme" method would look like, but I went with a fully explicit name: AddXyzSignInCookie.

```C#
        // Adds google auth and a default cookie auth for sign in purposes
        public static AuthenticationBuilder AddGoogleSignInCookie(this AuthenticationBuilder builder, Action<GoogleOptions> configureOptions)
        {
            builder.AddGoogle(GoogleDefaults.AuthenticationScheme, o =>
            {
                o.SignInScheme = CookieAuthenticationDefaults.AuthenticationScheme; // REVIEW: Should we override instead of default?
                configureOptions?.Invoke(o);
            });

            builder.Services.Configure<AuthenticationOptions>(o =>
            {
                if (!o.SchemeMap.ContainsKey(CookieAuthenticationDefaults.AuthenticationScheme))
                {
                    // TODO: make it easier to reuse AddCookie
                    builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<CookieAuthenticationOptions>, PostConfigureCookieAuthenticationOptions>());
                    builder.AddScheme<CookieAuthenticationOptions, CookieAuthenticationHandler>(CookieAuthenticationDefaults.AuthenticationScheme, CookieAuthenticationDefaults.AuthenticationScheme, _ => { });
                }
            });

            builder.AddVirtualScheme("GoogleSignIn", "GoogleSignIn", o =>
            {
                o.Default = CookieAuthenticationDefaults.AuthenticationScheme;
                o.Challenge = GoogleDefaults.AuthenticationScheme;
            });
            return builder;
        }

```

I verified with the social example that this works as expected when you target the new "GoogleSignIn" (virtual) scheme.  

So startup basically will go from:

old:
```
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
                .AddCookie()
                .AddGoogle(o => {
                  o.ClientId = Configuration["google:clientid"];
                  o.ClientSecret = Configuration["google:clientsecret"];
                })
```

New:
```C#
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
                .AddGoogleSignInCookie(o => {
                   o.ClientId = Configuration["google:clientid"];
                   o.ClientSecret = Configuration["google:clientsecret"];
                })
```

Also any additional AddFacebookSignInCookie calls would share the same sign in cookie.  This would dovetail nicely into https://github.com/aspnet/Identity/issues/1409 where we are trying to split identity up so there's some kind of `AddAuthentication().AddIdentitySignInCookies()`

Thoughts?

**Comment by HaoK** [comment]
Looks like we can't completely combine the VirtualSchemeHandler and normal AuthenticationHandler due to SignIn/SignOut not being implemented in the base AuthenticationHandler.

That said, we can still roll the VirtualSchemeOptions into the base AuthenticationSchemeOptions so all handlers will be able to forward authenticate/challenge/forbid using the new option settings.

But the over load for a 'pure virtual scheme', will need to add a different handler that adds support for sign in + sign out forwarding.

**Comment by johnkors** [comment]
Hi, @HaoK ! @Eilon mentioned that this issue here might solve the problems I mention in #1559, but I don't follow. Could you shed some light on the subject..?

**Comment by Tratcher** [comment]
@johnkors look at the currently checked in version of this feature:
https://github.com/aspnet/Security/blob/e69d9e20635abcc8ae46970f819397b717b2527a/src/Microsoft.AspNetCore.Authentication/VirtualAuthenticationHandler.cs
It provides dynamic scheme selection based on a DefaultSelector action. That code isn't using anything new from 2.1, so you could copy it as is into a 2.0 app.

**Comment by HaoK** [comment]
Forwarding was merged into base via c0b8be58ba0ccdb3a4cd38825c30536f60ab99ce

SignIn changes will be tracked via https://github.com/aspnet/Security/issues/1558
